api/genericresource/resource_management.go

```diff
@@ -0,0 +1,55 @@
+package genericresource
+
+import (
+	"github.com/docker/swarmkit/api"
+)
+
+// Claim assign GenericResources to the task by taking them from the node's GenericResource
+// and storing them in the taken store
+func Claim(nodeRes *api.GenericResource, taskRes *api.GenericResource, taken *api.GenericResource) error {
+	t, err := getGenericResourceInteger(taskRes)
+	if err != nil {
+		return err
+	}
+
+	switch nr := nodeRes.Resource.(type) {
+	case *api.GenericResource_Integer:
+		nr.Integer.Val -= t.Integer.Val
+		taken.Resource = NewInteger(t.Integer.Val)
+	case *api.GenericResource_Set:
+		i := uint64(0)
+		res := NewSet()
+
+		for k := range nr.Set.Val {
+			if i >= t.Integer.Val {
+				break
+			}
+
+			i++
+
+			res.Set.Val[k] = nil
+			delete(nr.Set.Val, k)
```

Comment:
Would it be useful to check if the requested key exists in `nr.Set.Val`, and return an error if it does not?